pyiceberg/table/__init__.py

```diff
@@ -517,9 +546,6 @@ def overwrite(
         if not isinstance(df, pa.Table):
             raise ValueError(f"Expected PyArrow table, got: {df}")
 
-        if overwrite_filter != AlwaysTrue():
-            raise NotImplementedError("Cannot overwrite a subset of a table")
-
         if len(self._table.spec().fields) > 0:
             raise ValueError("Cannot write to partitioned tables")
```

Comment:
I think we can write to partitioned tables now. Shall we replace this with
```python
        if unsupported_partitions := [
            field for field in self.table_metadata.spec().fields if not field.transform.supports_pyarrow_transform
        ]:
            raise ValueError(
                f"Not all partition types are supported for writes. Following partitions cannot be written using pyarrow: {unsupported_partitions}."
            )
```
I did a simple test for this and it works. :

```python
spark.sql(
    f"""CREATE TABLE {catalog_name}.pyiceberg_test.{table_name} (
                number_partitioned  int,
                number              int
            )
            USING iceberg
            PARTITIONED BY (number_partitioned)"""
)

spark.sql(
    f"""INSERT INTO {catalog_name}.pyiceberg_test.{table_name} VALUES (1, 2), (1, 3), (1, 4), (2, 1), (2, 2), (2, 3), (3, 1), (3, 2), (3, 3)"""
)

data = [
    {"number_partitioned": 1, "number": 100},
    {"number_partitioned": 2, "number": 100},
    {"number_partitioned": 3, "number": 100},
]
df = pa.Table.from_pylist(
    data, schema=pa.schema([("number_partitioned", pa.int32()), ("number", pa.int32())])
)

tbl.overwrite(df, overwrite_filter="number = 2")
print(tbl.scan().to_pandas())
===== output ====
   number_partitioned  number
0                   1     100
1                   2     100
2                   3     100
3                   1       3
4                   1       4
5                   2       1
6                   2       3
7                   3       1
8                   3       3
```

Reply:
Great catch, I think this was from before partitioned writes were supported 👍